### PR TITLE
refactors needed for integrating verkle tree 

### LINKED ISSUE #5329: refactor StateProvider and StorageProvider behind one interface IWorldState
With Verkle Trees, we dont have seperate state and storage tree and there is no concept of storage root. Everything is stored in a single tree

### PR DESCRIPTION
**Pull request by tanishqjasoria**
Resolves #5329 

some interface and API changes for integrating verkle trees with nethermind. 
Also, bring state access under limited interfaces across the solution.

## Changes

- use WorldState instead of State and Storage provider
- Use the interface IReadOnlyTxProcessorSource instead of ReadOnlyTxProcessingEnv
- merge the functions UpdateCode and UpdateCodeHash  in StateProvider
- create a interface ISyncTrieStore to check best last state using trie store

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a fix or a feature that would cause existing functionality not to work as expected)
- [ ] Documentation update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional or API changes)
- [ ] Build-related changes
- [x] Other: _description_ 

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [x] No

#### Notes on testing

the best way to test would be to run a sync and then verify the trie

### CODE REVIEW

File: src/Nethermind/Nethermind.Api/NethermindApi.cs [hunk 1]
```diff
@@ -75,12 +76,7 @@ public IBlockchainBridge CreateBlockchainBridge()
             LazyInitializer.EnsureInitialized(ref _readOnlyDbProvider, () => new ReadOnlyDbProvider(DbProvider, false));
 
             // TODO: reuse the same trie cache here
-            ReadOnlyTxProcessingEnv readOnlyTxProcessingEnv = new(
-                _readOnlyDbProvider,
-                ReadOnlyTrieStore,
-                readOnlyTree,
-                SpecProvider,
-                LogManager);
+            IReadOnlyTxProcessorSourceExt readOnlyTxProcessingEnv = new ReadOnlyTxProcessingEnv(_readOnlyDbProvider, ReadOnlyTrieStore, readOnlyTree, SpecProvider, LogManager);
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Api/NethermindApi.cs:79]: NP: Prefer to put long params in multiple line like before. Keep code to the left and in the same column, easier to read. Also, any addition/remove/change to a parameter will appear as a single line in diff viewer.

File: src/Nethermind/Nethermind.Blockchain.Test/BlockProcessorTests.cs [hunk 2]
```diff
@@ -41,15 +41,14 @@ public void Prepared_block_contains_author_field()
...
             TrieStore trieStore = new(stateDb, LimboLogs.Instance);
-            IStateProvider stateProvider = new StateProvider(trieStore, codeDb, LimboLogs.Instance);
+            IWorldState stateProvider = new WorldState(trieStore, codeDb, LimboLogs.Instance);
             ITransactionProcessor transactionProcessor = Substitute.For<ITransactionProcessor>();
             BlockProcessor processor = new(
                 RinkebySpecProvider.Instance,
                 TestBlockValidator.AlwaysValid,
                 NoBlockRewards.Instance,
-                new BlockProcessor.BlockValidationTransactionsExecutor(transactionProcessor, stateProvider),
+                new BlockProcessor.BlockValidationTransactionsExecutor(transactionProcessor, new WorldState(trieStore, codeDb, LimboLogs.Instance)),
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Blockchain.Test/BlockProcessorTests.cs:50]: State provider already in line 44?

File: src/Nethermind/Nethermind.Blockchain.Test/BlockProcessorTests.cs [hunk 3]
```diff
@@ -103,15 +100,14 @@ public void Recovers_state_on_cancel()
...
+            IWorldState worldState = new WorldState(trieStore, codeDb, LimboLogs.Instance);
             ITransactionProcessor transactionProcessor = Substitute.For<ITransactionProcessor>();
             BlockProcessor processor = new(
                 RinkebySpecProvider.Instance,
                 TestBlockValidator.AlwaysValid,
                 new RewardCalculator(MainnetSpecProvider.Instance),
-                new BlockProcessor.BlockValidationTransactionsExecutor(transactionProcessor, stateProvider),
-                stateProvider,
-                new StorageProvider(trieStore, stateProvider, LimboLogs.Instance),
+                new BlockProcessor.BlockValidationTransactionsExecutor(transactionProcessor, new WorldState(trieStore, codeDb, LimboLogs.Instance)),
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Blockchain.Test/BlockProcessorTests.cs:109]: Another double instantiation?

File: src/Nethermind/Nethermind.Blockchain.Test/Producers/DevBlockproducerTests.cs [hunk 4]
```diff
@@ -53,30 +53,24 @@ public void Test()
...
                 virtualMachine,
                 LimboLogs.Instance);
             BlockProcessor blockProcessor = new(
                 specProvider,
                 Always.Valid,
                 NoBlockRewards.Instance,
-                new BlockProcessor.BlockValidationTransactionsExecutor(txProcessor, stateProvider),
-                stateProvider,
-                storageProvider,
+                new BlockProcessor.BlockValidationTransactionsExecutor(txProcessor, new WorldState(trieStore, dbProvider.RegisteredDbs[DbNames.Code], LimboLogs.Instance)),
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Blockchain.Test/Producers/DevBlockproducerTests.cs:72]: Another double instantiation.

File: src/Nethermind/Nethermind.State/IStateProvider.cs [hunk 5]
```diff
@@ -22,7 +21,7 @@ public interface IStateProvider : IReadOnlyStateProvider, IJournal<int>
 
         void CreateAccount(Address address, in UInt256 balance, in UInt256 nonce);
 
-        void UpdateCodeHash(Address address, Keccak codeHash, IReleaseSpec spec, bool isGenesis = false);
+        void InsertCode(Address address, ReadOnlyMemory<byte> code, IReleaseSpec spec, bool isGenesis = false);
```
Comment (asdacap) [inline src/Nethermind/Nethermind.State/IStateProvider.cs:24]: NP: Recommend Span instead of memory for the interface. You can make a Span from Memory, but not the other way around.

File: src/Nethermind/Nethermind.State/IWorldState.cs [hunk 6]
```diff
@@ -1,22 +1,61 @@
...
-        IStorageProvider StorageProvider { get; }
+        new void RecalculateStateRoot();
 
-        IStateProvider StateProvider { get; }
+        new Keccak StateRoot { get; set; }
 
-        Snapshot TakeSnapshot(bool newTransactionStart = false);
+        new void DeleteAccount(Address address);
 
+        new void CreateAccount(Address address, in UInt256 balance);
```
Comment (asdacap) [inline src/Nethermind/Nethermind.State/IWorldState.cs:24]: What does `new` do here?

File: src/Nethermind/Nethermind.Synchronization.Test/SyncProgressResolverTests.cs [hunk 7]
```diff
@@ -57,11 +57,12 @@ public void Best_state_is_head_when_there_are_no_suggested_blocks()
             IBlockTree blockTree = Substitute.For<IBlockTree>();
             IReceiptStorage receiptStorage = Substitute.For<IReceiptStorage>();
             IDb stateDb = Substitute.For<IDb>();
+            ITrieStore trieStore = new TrieStore(stateDb, LimboLogs.Instance);
             SyncConfig syncConfig = new();
             syncConfig.PivotNumber = "1";
             ProgressTracker progressTracker = new(blockTree, stateDb, LimboLogs.Instance);
 
-            SyncProgressResolver syncProgressResolver = new(blockTree, receiptStorage, stateDb, NullTrieNodeResolver.Instance, progressTracker, syncConfig, LimboLogs.Instance);
+            SyncProgressResolver syncProgressResolver = new(blockTree, receiptStorage, stateDb, trieStore, progressTracker, syncConfig, LimboLogs.Instance);
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Synchronization.Test/SyncProgressResolverTests.cs:65]: This was null, is this change expected?

File: src/Nethermind/Nethermind.Trie/VisitContext.cs [hunk 8]
```diff
@@ -12,16 +13,31 @@ public class TrieVisitContext : IDisposable
         private readonly int _maxDegreeOfParallelism = 1;
         private int _visitedNodes;
 
-        public int Level { get; internal set; }
+        public int Level { get; set; }
         public bool IsStorage { get; internal set; }
         public int? BranchChildIndex { get; internal set; }
         public bool ExpectAccounts { get; init; }
         public int VisitedNodes => _visitedNodes;
+        public bool KeepTrackOfAbsolutePath { get; init; }
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Trie/VisitContext.cs:21]: Is this related to the PR?

File: src/Nethermind/Nethermind.Core/CodeChunks.cs [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Diagnostics;
+using Nethermind.Int256;
+
+namespace Nethermind.Core;
+
+[DebuggerDisplay("{Address}->{Chunk}")]
+public readonly struct CodeChunk : IEquatable<CodeChunk>
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Core/CodeChunks.cs:11]: Where is this used at?

**Comment by asdacap** [comment]
This is not based on 'master', is this intended?

**Comment by tanishqjasoria** [comment]
> This is not based on 'master', is this intended?
yes, this is not supposed to be merged right now - there is no potential usecase making this big a change. This change is mostly needed for verkle trees.
The reason I wanted to get this reviewed is to make sure that this direction is okay to move in as the verkle tree changes is very much dependent on this refactor.

**Comment by tanishqjasoria** [comment]
https://github.com/NethermindEth/nethermind/pull/5533
